modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/ObjectRelationship/ObjectRelationship.js

```diff
@@ -12,33 +12,101 @@
  * details.
  */
 
-import Autocomplete from 'commerce-frontend-js/components/autocomplete/Autocomplete';
+import {useIsMounted} from '@liferay/frontend-js-react-web';
 import {FieldBase} from 'dynamic-data-mapping-form-field-type/FieldBase/ReactFieldBase.es';
-import React from 'react';
+import {Autocomplete as BaseAutocomplete} from 'frontend-js-components-web';
+import {debounce, fetch} from 'frontend-js-web';
+import React, {useState} from 'react';
 
-const ObjectRelationship = ({
+function getData(apiUrl, query, page, pageSize) {
+	const url = new URL(apiUrl, themeDisplay.getPortalURL());
+
+	if (query) {
+		url.searchParams.set('search', query);
+	}
+
+	if (page) {
+		url.searchParams.set('page', page);
+	}
+
+	if (pageSize) {
+		url.searchParams.set('pageSize', pageSize);
+	}
+
+	return fetch(url, {
+		headers: new Headers({
+			Accept: 'application/json',
+			'Accept-Language': themeDisplay.getBCP47LanguageId(),
+			'Content-Type': 'application/json',
+		}),
+	}).then((data) => data.json());
+}
+
+export function ObjectRelationship({
 	apiURL,
 	initialLabel,
 	initialValue,
 	inputName,
-	itemsKey,
-	itemsLabel,
+	labelKey,
 	name,
 	value,
+	valueKey,
 	...otherProps
-}) => (
-	<FieldBase name={name} {...otherProps}>
-		<Autocomplete
-			apiUrl={apiURL}
-			initialLabel={initialLabel}
-			initialValue={initialValue}
-			inputName={inputName}
-			itemsKey={itemsKey}
-			itemsLabel={itemsLabel}
-			name={name}
-			value={value}
-		/>
-	</FieldBase>
-);
+}) {
+	const [query, setQuery] = useState(initialLabel || '');
+	const [selectedItem, setSelectedItem] = useState(initialValue || value);
+	const [items, setItems] = useState(null);
+	const [loading, setLoading] = useState(false);
+	const isMounted = useIsMounted();
+
+	const fetchData = debounce(() => {
+		if (isMounted()) {
+			setLoading(true);
+
+			getData(apiURL, query, 1, 10)
+				.then((jsonResponse) => {
+					setItems(jsonResponse.items);
+
+					setLoading(false);
+
+					if (!query) {
+						return;
+					}
+
+					const found = jsonResponse.items.find(
+						(item) => item[labelKey] === query
+					);
+
+					if (found) {
+						setSelectedItem(found);
+					}
+				})
+				.catch(() => {
+					setLoading(false);
+				});
+		}
+	}, 500);
```

Comment:
I think this could be replaced with `useResource`.

```js
import {fetch} from 'frontend-js-web';

const [networkStatus, setNetworkStatus] = useState(4);
const {resource: items} = useResource({
	fetch,
	fetchPolicy: 'cache-first',
	link: new URL(apiUrl, themeDisplay.getPortalURL()),
	onNetworkStatusChange: setNetworkStatus,
	variables: {
		search: query,
		page: 1,
		pageSize: 10
	}
});

const initialLoading = networkStatus === 1;
const loading = networkStatus < 4;
const error = networkStatus === 5;
```

Reply:
I ended up using this in one place but not the other. I didn't use it for the infinite scroll because I thought it would be odd to refetch and update the list in a `useEffect`.

Reply:
Hmm I think I saw this case in the ListView, maybe we can use, as I understand the rule, it calls `fetchData` if the value of `query` changes, in this case with `useResource` and the query remains available in `variables` the `useResource` will automatically request again and thus no longer need `useEffect` for this. In the `onBottomTouched` call it's also no longer necessary because you're going to be using `page` as part of `variables` too:

```js
const {resource: items} = useResource({
	...
	variables: {
		search: query,
		page,
		pageSize,
	}
});
``` 

Does it make sense? I can be missing something.

Reply:
My concern is that we have to still maintain a list of _all_ items requested since the list grows on scroll. So if we use `useResource`, we have to use an effect for updating the list. Unless there is another way to join the new requested items with the old requested items because we want to render page 1 + page 2 and not just render one or the other.

```js
const [listItems, setListItems] = useState([]);

const {resource: items} = useResource({
	...
	variables: {
		search: query,
		page,
		pageSize,
	}
});

useEffect(() => {
	setListItems([...listItems, ...items]);
}, [items])
```

If we rely on the useEffect, I think that would cause multiple re-renders, once when items is updated with page 2 and then again when we add that with setListItems

Reply:
Oh I see, now this makes more sense to me, in fact using `useResource` wouldn't help in this case, I'm going through the same situation in implementing Tree View for loading more items and maybe `useResource` will support this via OOTB, still drawing as the API would be...

But I agree. Maybe it could just remove the `fetchData` implementation and shorten the code but still having to use `useEffect` or `useMemo` can help too.